types/number.go

```diff
@@ -1,11 +1,82 @@
 package types
 
-import "strconv"
+import (
+	"math"
+	"strconv"
+)
 
 type (
 	Number float64
 )
 
+var ε = math.Nextafter(1, 2) - 1
+
+func NewNumber(a float64) Number {
+	return Number(a)
+}
+
+func (a Number) Value() float64 { return float64(a) }
+
 func (a Number) String() string {
 	return strconv.FormatFloat(float64(a), 'f', -1, 64)
-}
\ No newline at end of file
+}
+
+// https://es5.github.io/#x9.2
+func (a Number) IsTrue() bool {
+	return !a.IsFalse()
+}
+
+func (a Number) IsFalse() bool {
+	return math.IsNaN(float64(a)) ||
+		a == -0.0 ||
+		a == +0.0
+}
+
+// ToBool returns a Boolean according to:
+// https://es5.github.io/#x9.2
+func (a Number) ToBool() Bool {
+	if a.IsTrue() {
+		return Bool(true)
+	}
+
+	return Bool(false)
+}
+
+// ToNumber retrieves the number.
+// https://es5.github.io/#x9.3
+func (a Number) ToNumber() Number {
+	return a
+}
+
+// ToString converts the number to string.
+// Check https://es5.github.io/#x9.8
+// TODO(i4k): revisit this.
+func (a Number) ToString() String {
+	val := strconv.FormatFloat(float64(a), 'f', -1, 64)
+	return NewString(val)
+}
+
+func (_ Number) Kind() Kind {
+	return KindNumber
+}
+
+func (a Number) Equal(b Number) bool {
+	if math.IsNaN(a.Value()) ||
```

Comment:
NaN != NaN ? o.O

Reply:
hahaha Good question. Probably this has a bug.

Reply:
I'll take a look.

Reply:
hmm if it's not a number.. then it's not comparable.

Reply:
Take a look in the spec:

http://es5.github.io/#x11.9.6